backend/selectgen.ml

```diff
@@ -773,23 +798,52 @@ method insert_op_debug env op dbg rs rd =
 method insert_op env op rs rd =
   self#insert_op_debug env op Debuginfo.none rs rd
 
-(* Add the instructions for the given expression
-   at the end of the self sequence *)
-
+method insert_endregions env regions =
+  match regions with
+  | [] -> ()
+  | regions ->
+     (* Coalesce multiple simultaneous Iendregion *)
+     let final_region = List.hd (List.rev regions) in
+     self#insert env (Iop Iendregion) final_region [| |]
+
+method private maybe_endregion env region =
+  match region with
+  | None -> ()
+  | Some reg -> self#insert_endregions env [reg]
+
+(* Emit an expression, which is assumed not to end any regions early.
+   (This holds for any expression not in tail position of Cregion)
+
+   Returns:
+     - [None] if the expression does not finish normally (e.g. raises)
+     - [Some rs] if the expression yields a result in registers [rs] *)
 method emit_expr (env:environment) exp =
-  (* Environment used in recursive calls not in tail position *)
-  let env' =
-    if env.region_tail then {env with region_tail=false} else env in
+  match self#emit_expr_aux env exp with
+  | None -> None
+  | Some (res, unclosed) ->
+     assert (unclosed == env.regions);
+     Some res
+
+(* Emit an expression which may end some regions early.
+
+   Returns:
+    - [None] if the expression does not finish normally (e.g. raises)
+    - [Some (rs, unclosed)] if the expression yields a result in [rs],
+      having left [unclosed] (a suffix of env.regions) regions open *)
+method emit_expr_aux (env:environment) exp :
```

Comment:
It's not really in the interface, though: because of classes, that `mli` contains a bunch of implementation details:
```
  (* The following methods should not be overridden.  They cannot be
     declared "private" in the current implementation because they
     are not always applied to "self", but ideally they should be private. *)
```

Reply:
Fair. I'd still like a more descriptive name than `aux`. Since `emit_tail` is taken, how about: `emit_expr_region_tail` or `emit_expr_trailing`?

Reply:
I don't like `emit_expr_region_tail` because you can reach this code in non-tail position of a region. (I'd rather the name be vague than be wrong: vague makes people read comments and wrong makes people make mistakes).

`emit_expr_trailing` is vague enough to work, though! (I struggled for a while to pick a name that was neither vague nor wrong, but didn't come up with much)